ArC: initial support for inactive synthetic beans

Ladicek opened the pull request:
Fixes #41466

Ladicek commented:
This PR is work in progress, hence the draft status. I implemented some (but not all) ideas from #41466 (mostly because I didn't really understand it all and I am not sure how it all fits together) and would like @yrodiere to take it up for a spin before declaring this PR as ready.

Ladicek commented:
Just to be sure: the synthetic bean builder (`BeanConfigurator`, `ExtendedBeanConfigurator`) gained one new method, `checkActive()`, to configure the active check function (which is a `Runnable`). There already is a method to force eager initialization, it's called `startup()`, so I didn't feel compelled to add another. Eager initialization is skipped for beans that are determined to be inactive.

yrodiere commented:
Hey @Ladicek , thanks for working on this.

I'll try to integrate this ASAP, but in the meantime, a few comments:

> There already is a method to force eager initialization, it's called `startup()`, so I didn't feel compelled to add another. Eager initialization is skipped for beans that are determined to be inactive.

Okay, but the requirement was to eagerly initialize only if the bean is active *and* injected into a user bean. From the issue:

> We want startup to fail if the application contains a user bean that gets injected with an active , unconfigured (no JDBC URL) datasource.

I'll have to try, but I suspect without this additional condition, we'll have a few problems, like users discovering only at runtime that their datasource (which they inject and use) didn't start because of missing configuration.

---

Relatedly, one of the main purposes of implementing this in Arc itself (as opposed to doing it by hand in extensions) was this:

> We want the failures to include actionable messages:
>   * The root cause for the problem, specific to each bean: "This datasource is inactive because quarkus.datasource.active is set to false" (**DONE**)
>     * For injected beans, **the list of injection points** (**NOT DONE**)

---

I will let you know of my results when I've had a try. Thanks again.

Ladicek commented:
> Okay, but the requirement was to eagerly initialize only if the bean is active _and_ injected into a user bean. From the issue:
> 
> > We want startup to fail if the application contains a user bean that gets injected with an active , unconfigured (no JDBC URL) datasource.
> 
> I'll have to try, but I suspect without this additional condition, we'll have a few problems, like users discovering only at runtime that their datasource (which they inject and use) didn't start because of missing configuration.

Yeah, so that's one of the things I didn't understand. How can you have an _active_ but _unconfigured_ datasource? If the datasource is unconfigured, it is inactive, like, by definition. No?

I suppose this is an area where more work is required, but I don't really understand what work that is, so I left it for later.

> Relatedly, one of the main purposes of implementing this in Arc itself (as opposed to doing it by hand in extensions) was this:
> 
> > We want the failures to include actionable messages:
> > 
> > * The root cause for the problem, specific to each bean: "This datasource is inactive because quarkus.datasource.active is set to false" (**DONE**)
> >   
> >   * For injected beans, **the list of injection points** (**NOT DONE**)

Actually, also **DONE**. At least if I understand the requirement correctly. If there's an inactive bean, any attempt to create an instance of that bean (which will happen during startup if the synthetic bean is configured with `startup()`) will fail with an error whose message includes:

- the message from the exception thrown by `checkActive()`
- the list of injection points that resolve to that [inactive] bean, if there are any

yrodiere commented:
> Yeah, so that's one of the things I didn't understand. How can you have an _active_ but _unconfigured_ datasource? If the datasource is unconfigured, it is inactive, like, by definition. No?

I think that's how we ended up defining this... though that's not the *current* definition. I have a hard time determining if there are cases where we actually want an unconfigured datasource to be considered active (e.g. to get specific errors). I'll I'll have to try.

That being said... if we accept your definition, the requirement becomes this:

> We want startup to fail if the application contains a user bean that gets injected with an inactive datasource.

I.e.: since we're now skipping eager initialization for datasources that are inactive, we need to check that no other (user) bean is using these datasources, otherwise the user may start their app forgetting to set the JDBC URL, think everything is fine, and see the app fail on the first request (first time the datasource bean gets accessed).

Ladicek commented:
Ouch, yeah, good point. If the bean is not active, initialization on `startup()` shall be skipped, which is not what we want _sometimes_ (currently: when the bean is injected into a non-synth bean). I'll have to fix that. Thanks for pointing that out!

mkouba commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InjectableBean.java:
```diff
@@ -162,6 +162,13 @@ default Class<?> getImplementationClass() {
+     * Checks whether this (synthetic) bean is active. To signal that this bean is <em>not</em> active,
+     * this method should throw an exception. If this method does not throw, this bean is considered active.
+     */
+    default void checkActive() {
```
Why do we need this method? Could we actually do the check when a bean is created?

Ladicek replied:
Good point, we can generate the bytecode directly instead of going through this method. Will remove.

Ladicek replied:
Done.

mkouba commented on extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/SyntheticBeanBuildItem.java:
```diff
@@ -181,6 +192,19 @@ public ExtendedBeanConfigurator setRuntimeInit() {
+         * @return self
+         * @throws IllegalArgumentException If the {@code checkActive} argument is not a proxy returned from a recorder method
+         */
+        public ExtendedBeanConfigurator checkActive(Runnable checkActive) {
```
It might make sense to add a variant that would accept a bytecode generator, similarly to `io.quarkus.arc.processor.BeanConfiguratorBase.creator(Consumer<MethodCreator>)`.

Ladicek replied:
There's `BeanConfiguratorBase.checkActive(Consumer<MethodCreator>)`, so I think that's already present.

Ladicek commented:
> Ouch, yeah, good point. If the bean is not active, initialization on `startup()` shall be skipped, which is not what we want _sometimes_ (currently: when the bean is injected into a non-synth bean). I'll have to fix that. Thanks for pointing that out!

Done.

yrodiere commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InactiveBeanException.java:
```diff
@@ -0,0 +1,12 @@
+/**
+ * Indicates that an inactive bean was injected (or dynamically looked up).
+ */
+public class InactiveBeanException extends InjectionException {
```
So first, having this exception is great, and allows me to do everything I need.

That being said, I would like to point out two things:

* Right now the only way to check if a bean is inactive is [to fetch it](https://github.com/quarkusio/quarkus/pull/41929/files#diff-93d12c860e8b94edea17693d4d0ac031cef765d2c74cc758f283839e4ae00af9R33-R38). Do we want to introduce `InjectableInstance#isActive()` to allow easy programmatic lookup?
* Right now the only way to retrieve a bean optionally, only if it's active, is to [catch the exception and return null/optional](https://github.com/quarkusio/quarkus/pull/41929/files#diff-93d12c860e8b94edea17693d4d0ac031cef765d2c74cc758f283839e4ae00af9R25-R31). And when retrieving multiple beans (e.g. all instances of `AgroalDataSource`) that can get rather impractical; one would most likely have to use `InjectableInstance#handles`. Do we want to add an `Active` qualifier that could be passed to `BeanContainer#select()` to tell Arc to ignore inactive beans?

Note I'm not suggesting to to anything more in this PR, it can definitely be a follow-up issue -- which may even die a slow death :)

Ladicek replied:
I don't think an `@Active` qualifier is a good idea, because all other injection points would suddenly also have to use it.

Previously, I had a method `InjectableBean.checkActive()`, which I deleted but could add again. Perhaps in a slightly altered form: `boolean isActive()`. An `InjectableBean` is available from `InstanceHandle`. That seems rather straightforward to me:

```java
@Inject
@Any
Instance<AgroalDataSource> ds;

...

for (InstanceHandle handle : ds.handles()) {
    if (handle.getBean().isActive()) {
        AgroalDataSource datasource = handle.get();
        ...
    }
}
```

Doesn't look all that bad to me?

yrodiere replied:
> Previously, I had a method `InjectableBean.checkActive()`, which I deleted but could add again. Perhaps in a slightly altered form: `boolean isActive()`.

LGTM

> I don't think an @Active qualifier is a good idea, because all other injection points would suddenly also have to use it.

I suppose you're saying that because the bean would (per spec) lose the `@Default` qualifier?

Since `@Active` is not in the spec, I'd say you're free to declare that it has special behavior, e.g. it can't be applied to a bean explicitly, and it doesn't lead to the `@Default` qualifier being dropped.

After all, there's precedent with `@Any`: that qualifier is added automatically on every bean, and doesn't affect `@Default`.

> Doesn't look all that bad to me?

It's less practical than a one-liner, but I agree it's reasonable. Probably the best compromise we can get if the `@Active` qualifier doesn't make sense.

Well, I guess a `Instance#activeHandles()` method would be one step closer to what I suggested, but it feels wonky.

Ladicek replied:
> I suppose you're saying that because the bean would (per spec) lose the `@Default` qualifier?

Yeah, that's what I meant.

It is true that we could special-case the `@Active` qualifier (like `@Any` and `@Named` are, and to some extent our `@All`), but I'd like to avoid that just because special cases in this are are super surprising (cf. `@Named`).

It does have certain appeal, though. I'll think some more about it.

yrodiere replied:
FWIW, everything works without that `@Active` "pseudo-qualifier" in my PR, and it works fine in extension code. But I must admit it feels a bit complicated when we suggest it to application developers.

See for example [here](https://quarkus-pr-main-41929-preview.surge.sh/version/main/guides/datasource#datasource-active), in the CDI producer definition:

```java
public class MyProducer {
    @Inject
    @DataSource("pg")
    InjectableInstance<AgroalDataSource> pgDataSourceBean; 

    @Inject
    @DataSource("oracle")
    InjectableInstance<AgroalDataSource> oracleDataSourceBean;

    @Produces 
    @ApplicationScoped
    public AgroalDataSource dataSource() {
        if (pgDataSourceBean.getHandle().getBean().isActive().result()) { 
            return pgDataSourceBean.get();
        } else if (oracleDataSourceBean.getHandle().getBean().isActive().result()) { 
            return oracleDataSourceBean.get();
        } else {
            throw new RuntimeException("No active datasource!");
        }
    }
}
```

I suppose we could say it's "not great, not terrible" :)

But with an `@Active` qualifier, and if that qualifier could be made to work in static injection points as well... we could get simpler code:

```java
public class MyProducer {
    @Inject
    @DataSource("pg")
    @Active
    Instance<AgroalDataSource> pgDataSourceBean; 

    @Inject
    @DataSource("oracle")
    @Active
    Instance<AgroalDataSource> oracleDataSourceBean;

    @Produces 
    @ApplicationScoped
    public AgroalDataSource dataSource() {
        if (pgDataSourceBean.isResolvable()) { 
            return pgDataSourceBean.get();
        } else if (oracleDataSourceBean.isResolvable()) { 
            return oracleDataSourceBean.get();
        } else {
            throw new RuntimeException("No active datasource!");
        }
    }
}
```

Or even just (though I'm not entirely sure about this one -- I suspect static injection combined with `@Active` may not be straightforward to implement):

```java
public class MyProducer {
    @Inject
    @Active
    AgroalDataSource activeDataSource;

    @Produces 
    @ApplicationScoped
    public AgroalDataSource dataSource() {
        return activeDataSource;
    }
}
```

Ladicek replied:
> I suppose we could say it's "not great, not terrible" :)

Agreed :-)

The `@Active` qualifier is appealing, but I'd rather leave it out of the initial iteration. We should be able to add it later.

yrodiere reviewed (commented):
Thanks!

I this gave a first try here (very incomplete, but it's a start): https://github.com/quarkusio/quarkus/pull/41929

I added comments below based on what I noticed.

Right now I'm kind of stuck because whenever there's a bean initialized on startup, and it's inactive, startup fails... even if the bean is not injected in a user bean. See my comment below.

Ladicek commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java:
```diff
@@ -1089,6 +1090,41 @@ private void implementCreateForSyntheticBean(ClassCreator beanCreator, BeanInfo
+
+            TryBlock tryBlock = doCreate.tryBlock();
+            tryBlock.invokeSpecialMethod(checkActiveMethod.getMethodDescriptor(), tryBlock.getThis());
+            CatchBlockCreator catchBlock = tryBlock.addCatch(Exception.class);
```
I talked about this with @mkouba last week and we eventually concluded that a `Runnable` throwing an exception is a bad API. The proper one should be something like `Supplier<ActiveResult>`, where

```java
public final class ActiveResult {
    private final boolean active;
    private final String inactiveReason;

    public static ActiveResult active() {
        return new ActiveResult(true, null);
    }

    public static ActiveResult inactive(String reason) {
        return new ActiveResult(false, Objects.requireNonNull(reason));
    }

    private ActiveResult(boolean active, String inactiveReason) {
        this.active = active;
        this.inactiveReason = inactiveReason;
    }

    public boolean isActive() {
        return active;
    }

    public String inactiveReason() {
        return inactiveReason;
    }
}
```

If the supplier throws, we won't catch that and everything will blow up, which is probably better than what we have here.

yrodiere replied:
Can we add a cause into this? It's possible one bean would be inactive because another is inactive, e.g. Liquibase only makes sense if the relevant datasource is active, and in that case you'd want to forward the `InactiveBeanException` thrown by accessing the datasource.

Ladicek replied:
That sounds reasonable. I'm just wondering if the `cause` should be arbitrary `Exception`, or just `InactiveBeanException`...

yrodiere replied:
Either way works I suppose, as long as the cause's stack trace is displayed (so that users can see the cause of the cause).

yrodiere replied:
Hey @Ladicek,

So I've got some more input here... With the current code I ended up doing this:

```java
    public Supplier<ActiveResult> flywayActiveSupplier(String dataSourceName) {
        return new Supplier<ActiveResult>() {
            @Override
            public ActiveResult get() {
                FlywayDataSourceRuntimeConfig dataSourceConfig = config.getValue()
                        .getConfigForDataSourceName(dataSourceName);

                // Flyway can only be active when the datasource itself is active.
                try {
                    ClientProxy.unwrap(AgroalDataSourceUtil.dataSourceInstance(dataSourceName).get());
                } catch (InactiveBeanException e) {
                    if (dataSourceConfig.active.isPresent() && dataSourceConfig.active.get()) {
                        String propertyKey = FlywayContainerUtil.getFlywayPropertyKey(dataSourceName, "active");
                        throw new ConfigurationException(String.format(Locale.ROOT,
                                "Flyway for datasource '%1$s' was activated explicitly through configuration property '%2$s',"
                                        + " but datasource '%1$s' was deactivated."
                                        + " Either deactivate Flyway, or activate the datasource.",
                                dataSourceName, propertyKey),
                                Set.of(propertyKey));
                    }
                    return ActiveResult.inactive(String.format(Locale.ROOT,
                            "Flyway for datasource '%s' was deactivated automatically because this datasource was deactivated."
                                    + e.getMessage(),
                            dataSourceName));
                }

                // Even if the datasource is active, Flyway can still be deactivated explicitly
                if (dataSourceConfig.active.isPresent() && !dataSourceConfig.active.get()) {
                    return ActiveResult.inactive(String.format(Locale.ROOT,
                            "Flyway for datasource '%s' was deactivated through configuration properties."
                                    + " To keep Flyway deactivated and avoid this error:"
                                    + " inject Flyway-dependent CDI beans using Instance<Sometype>' instead of 'Sometype',"
                                    + " and make sure code only accesses these beans if Flyway actually is active."
                                    + " To activate Flyway: set configuration property '%s' to 'true'.",
                            dataSourceName, FlywayContainerUtil.getFlywayPropertyKey(dataSourceName, "active")));
                }

                return ActiveResult.active();
            }
        };
    }
```

As you can see, I need to detect that the datasource is inactive, because in that case Flyway must be deactivated, and to do that I must to try to retrieve the datasource, because that's the only way I can forward to the user the reason for the datasource (and Flyway) being deactivated.

That's the best I can do right now, but it's still not ideal, because the inactive reason for Flyway would include the full message of the `InactiveBeanException` for the datasource... including among other the list of beans that get injected with a `DataSource`, which is really not relevant when the problem is the injection point of a `Flyway` bean.

There's also the problem that the "inactive reason" for a datasource needs to be actionable, so in that reason I had to include advice such as "make sure you inject the datasource as `Instance<DataSource>` instead of just `DataSource`"... which really isn't relevant in the "inactive reason" of a `Flyway` bean -- errors involving that bean would not involve the injection of a `DataSource`.

To solve the above, I would suggest:

1. Changing `ActiveResult.inactive(String reason)` to `ActiveResult.inactive(Exception cause)`.
2. Preserving the existing way to build the `InactiveBeanException` message, just replacing the `.append(reason)` with `.append(cause.getMessage())`.
3. Adding more hints regarding how to avoid the `InactiveBeanException` in the `InactiveBeanException` message. See my other comment.
4. Always assigning the cause to `InactiveBeanException`.

With 1, 2 and 3, we make sure that users getting an `InactiveBeanException` will get all the information they need to get out of this situation -- with an actionable message, and a detailed stacktrace should they need it.
With 3, we make sure that extensions declaring inactive beans don't need to include information in their message that is irrelevant to consumers of "transitively deactivated" beans (e.g. Flyway)
With 4, we make sure that extensions (e.g. Flyway) trying to propagate the reason that another bean (e.g. DataSource) is inactive will have access to the full message, without any irrelevant information: they just need to call `inactiveBeanException.getCause().getMessage()`.

Ladicek replied:
It really feels like an inactive `ActiveReason` should be able to hold a cause, which would be another inactive `ActiveReason` (potentially recursively). We could then print the whole cause chain in the `InactiveBeanException` message.

To be able to access the causing `ActiveReason`, I ended up changing `InjectableBean.isActive()` from returning `boolean` to returning `ActiveResult`.

You should be able to do something like:

```java
ActiveResult active = AgroalDataSourceUtil.dataSourceInstance(dataSourceName).getHandle().getBean().isActive();
```

And then pass `active` to any `ActiveResult.inactive()` calls you make. The `InactiveBeanException` message then looks like:

```
Bean is not active: ...
Reason: top-level reason
Cause: first-level cause reason
Cause: second-level cause reason
...
```

WDYT? Does that make sense, or is exception chaining really better?

yrodiere replied:
Works for me, thanks! I'll try that.

You may want to change `isActive` to `getActive`, since the convention is usually to return a `boolean` from a `is*` method, but I don't _personally_ mind.

I could see exception chaining being better to display a proper stack trace, especially if a bean's inactivity is caused by a user exception for some reason. But right now I don't have such case, so `ActiveResult` chaining sounds good.

yrodiere replied:
FWIW, this works perfectly in my WIP PR, though you may want to have a look at my comments:

> You may want to change isActive to getActive, since the convention is usually to return a boolean from a is* method, but I don't personally mind.
> 
> I could see exception chaining being better to display a proper stack trace, especially if a bean's inactivity is caused by a user exception for some reason. But right now I don't have such case, so ActiveResult chaining sounds good.

Up to you!

Ladicek replied:
I've renamed the `ActiveResult`-returning method `isActive()` to `checkActive()`. I've also added a `boolean`-returning method `isActive()`, which does just `return checkActive().value();`.

Which reminds me that I've also renamed `ActiveResult.result()` to `value()`, just for better readability.

I've left out exception chaining. I guess that's something we might add in the future, if the need comes.

Ladicek commented on extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/SyntheticBeanBuildItem.java:
```diff
@@ -79,6 +84,8 @@ public static class ExtendedBeanConfigurator extends BeanConfiguratorBase<Extend
         private Function<SyntheticCreationalContext<?>, ?> fun;
         private boolean staticInit;
 
+        private Runnable checkActive;
```
It would be possible to pass `SyntheticCreationalContext`, I think. Are the dependencies registered as injection points (and therefore available from `SyntheticCreationalContext.getInjectedReference()`)?

Other than that, `checkActive` is called during `Bean.create()`, which means ArC should already have been booted up, so `Arc.cotnainer()` should be safe to use. But I'm not 100% sure, so don't quote me on that :-)

yrodiere replied:
> It would be possible to pass SyntheticCreationalContext, I think. Are the dependencies registered as injection points (and therefore available from SyntheticCreationalContext.getInjectedReference())?

In my case they are, yes. It seems like a reasonable requirement as far as I'm concerned.

Ladicek replied:
OK, I'll try to add this.

Ladicek replied:
I'm looking into this and I'm wondering, when you say

> The check might rely on other (startup-initialized) beans

does that mean you need to call methods on the other beans, or you need to check that the other beans are also active? Or both?

If you need to check that the other beans are also active, then using a `SyntheticCreationalContext` won't work, at least not always. The SCC contains contextual references for the injected beans, which means contextual instances for pseudo-scoped beans. If any of those is not active, we'll throw `InactiveBeanException` when creating the SCC. If the bean is normal-scoped, the contextual reference is a client proxy, so we'll only throw `InactiveBeanException` when calling a method, which will be in the active check procedure, so bearable, but probably not very useful still.

yrodiere replied:
> does that mean you need to call methods on the other beans, or you need to check that the other beans are also active? Or both?

In my case I just need to check that other beans are active. I could see calling methods on other beans being useful, but I don't have a use case for that right now.

> If you need to check that the other beans are also active, then using a SyntheticCreationalContext won't work, at least not always. The SCC contains contextual references for the injected beans, which means contextual instances for pseudo-scoped beans.

The beans I'm interested in are application-scoped.

> If any of those is not active, we'll throw InactiveBeanException when creating the SCC.

Uh, I didn't anticipate that. Aren't synthetic consumers supposed to be able to work even if their dependencies are inactive?

> If the bean is normal-scoped, the contextual reference is a client proxy, so we'll only throw InactiveBeanException when calling a method, which will be in the active check procedure, so bearable, but probably not very useful still.

I suppose I can use `ClientProxy.unwrap()`? But yeah I agree, it's a bit clunky.

Anyway... If I understand correctly, it would be a better idea to use `Arc.container()` to retrieve beans in `checkActive` and then [any of the ways we discussed for checking if beans are active](https://github.com/quarkusio/quarkus/pull/41810#discussion_r1679632971) instead?... right? In which case maybe exposing `SyntheticCreationalContext` there is not a good idea.

To be honest I was suggesting to expose `SyntheticCreationalContext` because I thought this was a "better" way to retrieve beans, but I have no idea why it's better. So maybe it isn't, at least in this case.

Ladicek replied:
> Aren't synthetic consumers supposed to be able to work even if their dependencies are inactive?

I guess that really depends on what the consumer does. The SCC contains [eagerly initialized] contextual references, so that cannot work in case a pseudo-scoped bean might be inactive. It would work for normal-scoped beans, but as we agreed, it's a bit weird. See below for a possibly better idea.

> I suppose I can use `ClientProxy.unwrap()`? But yeah I agree, it's a bit clunky.

Yeah, that's correct, `ClientProxy.unwrap()` would work.

> To be honest I was suggesting to expose `SyntheticCreationalContext` because I thought this was a "better" way to retrieve beans, but I have no idea why it's better. So maybe it isn't, at least in this case.

I guess that's true in this case. SCC gives you a contextual reference to a bean, while you'd need a `Bean` or `InjectableInstance` or something like that. So `Arc.container()` seems better, at least until we figure out a better replacement.

Maybe we could do something like

```java
interface ActiveCheckContext {
    <T> InstanceHandle<T> getInjectedBean(Class<T> requiredType, Annotation... qualifiers);
    <T> InstanceHandle<T> getInjectedBean(TypeLiteral<T> requiredType, Annotation... qualifiers);
}
```

We would reuse the injection points configured via `BeanConfigurator.addInjectionPoint()`, but instead of SCC that gives you contextual references, we'd give you lazily initialized `InstanceHandle`. From there, you could obtain the `InjectableBean` to call `isActive()`, or perhaps obtain the contextual reference to call methods.

yrodiere replied:
FWIW using `Arc.container()` to retrieve beans works perfectly fine in my WIP PR; marking this thread as resolved.

Ladicek commented on extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/StartupBuildSteps.java:
```diff
@@ -175,44 +178,67 @@ private void registerStartupObserver(ObserverRegistrationPhaseBuildItem observer
+                    }
+                }
+
+                if (!isInjectedInUserBean) {
+                    TryBlock tryBlock = bytecode.tryBlock();
+                    CatchBlockCreator catchBlock = tryBlock.addCatch(InactiveBeanException.class);
+                    catchBlock.returnVoid();
+                    bytecode = tryBlock;
+                }
```
Hmm, let me check.

Ladicek replied:
Here's a `git diff` that made `AgroalMetricsConfigActiveFalseTest` pass for me. The first change is an obvious bug I noticed and probably doesn't have anything to do with the test. The second change is what makes the test pass. If you want to check whether a bean exists, you actually have to make an attempt to create a contextual instance. If the bean is pseudo-scoped, calling `Instance.get()` is enough, but if the bean is normal-scoped, you'll only get a client proxy. `ClientProxy.unwrap()` is the easiest way to obtain the actual instance. (Calling `toString()` would also work, but is useless.)

```diff
diff --git a/extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalDataSourceBuildUtil.java b/extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalDataSourceBuildUtil.java
index cbc9d6d9df0..b3324a4118a 100644
--- a/extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalDataSourceBuildUtil.java
+++ b/extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalDataSourceBuildUtil.java
@@ -15,7 +15,7 @@ public static AnnotationInstance qualifier(String dataSourceName) {
         if (dataSourceName == null || DataSourceUtil.isDefault(dataSourceName)) {
             return AnnotationInstance.builder(Default.class).build();
         } else {
-            return AnnotationInstance.builder(DataSource.class).add("name", dataSourceName).build();
+            return AnnotationInstance.builder(DataSource.class).value(dataSourceName).build();
         }
     }
 }
diff --git a/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AgroalDataSourceUtil.java b/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AgroalDataSourceUtil.java
index 4eec9d94d8d..044ad5ad033 100644
--- a/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AgroalDataSourceUtil.java
+++ b/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AgroalDataSourceUtil.java
@@ -10,6 +10,7 @@
 import io.agroal.api.AgroalDataSource;
 import io.quarkus.agroal.DataSource;
 import io.quarkus.arc.Arc;
+import io.quarkus.arc.ClientProxy;
 import io.quarkus.arc.InactiveBeanException;
 import io.quarkus.arc.InjectableInstance;
 import io.quarkus.datasource.common.runtime.DataSourceUtil;
@@ -24,7 +25,9 @@ public static InjectableInstance<AgroalDataSource> dataSourceInstance(String dat
 
     public static Optional<AgroalDataSource> dataSourceIfActive(String dataSourceName) {
         try {
-            return Optional.of(dataSourceInstance(dataSourceName).get());
+            AgroalDataSource instance = dataSourceInstance(dataSourceName).get();
+            ClientProxy.unwrap(instance);
+            return Optional.of(instance);
         } catch (InactiveBeanException e) {
             return Optional.empty();
         }
```

yrodiere replied:
Thank you @Ladicek , and sorry about that oversight. I'll resume my work.

Ladicek committed:
ArC: initial support for inactive synthetic beans

yrodiere commented on extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/StartupBuildSteps.java:
```diff
@@ -175,44 +178,67 @@ private void registerStartupObserver(ObserverRegistrationPhaseBuildItem observer
+                boolean isInjectedInUserBean = false;
+                for (InjectionPointInfo ip : observerRegistration.getBeanProcessor().getBeanDeployment().getInjectionPoints()) {
+                    BeanInfo resolvedBean = ip.getResolvedBean();
+                    if (resolvedBean != null && !resolvedBean.isSynthetic() && resolvedBean.equals(bean)) {
```
I noticed injecting a deactivated datasource into a user bean didn't fail startup, so I looked at this condition and... I think the condition `resulvedBean.isSynthetic()` is wrong here?

If `resolvedBean.equals(bean)`, then the resolved bean is the bean with a `checkActive` method, so it is necessarily synthetic (only way to add a `checkActive` method right now). So `!resolvedBean.isSynthetic() && resolvedBean.equals(bean)` will always be `false`.

I think we want to check that the bean *being injected with `resolvedBean`* is non-synthetic, right? So the condition should rather be `!ip.getTargetBean().map(BeanInfo::isSynthetic).orElse(false)`. Or something like that: I'm not sure what we want to do when there is no target bean, because what does that even mean? :thinking:

Ladicek replied:
Ouch. Good point, I guess I didn't think too much about this. You're right, the `isSynthetic()` method should be called on the _target_ bean of the IP (which is the bean that declares the IP), not on the resolved bean (which is the bean the IP resolves to).

Ladicek replied:
Done.

Ladicek commented:
Rebased on current `main`, changed the active check procedure from `Runnable` to `Supplier<ActiveResult>` and fixed the problem in `StartupBuildSteps`.

yrodiere commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java:
```diff
@@ -1089,6 +1091,46 @@ private void implementCreateForSyntheticBean(ClassCreator beanCreator, BeanInfo
                 .setModifiers(ACC_PRIVATE);
         bean.getCreatorConsumer().accept(createSynthetic);
 
+        // TODO move below, after `SyntheticCreationalContextImpl` is created, and pass it to this method
```
Adding a comment here so we don't forget to handle the TODO.

Ladicek replied:
FTR, this TODO is wrong and I'm removing it from the code. This is related to the other thread, where we discussed whether the active check procedure should be `Function<SyntheticCreationalContext, ActiveResult>` and ended up deciding that it shouldn't. (It might end up being `Function<ActiveCheckContext, ActiveResult>` instead, but the TODO still wouldn't apply.)

yrodiere commented on extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/SyntheticBeansProcessor.java:
```diff
@@ -118,6 +127,25 @@ public void accept(MethodCreator m) {
+                ResultHandle runnable = mc.invokeInterfaceMethod(
+                        MethodDescriptor.ofMethod(Map.class, "get", Object.class, Object.class),
+                        staticMap, mc.load(name));
+                // TODO different message, possibly even different exception
```
Adding a comment here so we don't forget to handle the TODO.

yrodiere commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java:
```diff
@@ -1089,6 +1091,46 @@ private void implementCreateForSyntheticBean(ClassCreator beanCreator, BeanInfo
+            msg.append("\nReason: ");
+            msg.append(notActive.invokeVirtualMethod(
+                    MethodDescriptor.ofMethod(ActiveResult.class, "inactiveReason", String.class), activeResult));
+            if (!matchingIPs.isEmpty()) {
+                msg.append("\nThis bean is injected into:");
```
Could we add the following hint, so that providers of inactive beans (Agroal, reactive clients, Flyway, Liquibase, ...) don't need to add it in their own messages?

See my other message for a more detailed motivation.

```suggestion
            ResultHandle typeNameString = /* ... TODO ... */;
            msg.append("\nTo keep this bean inactive and avoid this exception:");
            msg.append("\n\t- Configure all extensions consuming this bean as inactive as well, if they allow it, e.g. 'quarkus.someextension.active = false'");
            msg.append("\n\t- Make sure that custom code accessing this bean only does so if the bean is active");
            if (!matchingIPs.isEmpty()) {
                msg.append("\n\t- Inject custom CDI beans with 'Instance<").append(typeNameString).append(">' instead of '").append(typeNameString).append("'");
                msg.append("\nThis bean is injected into:");
```

Ladicek replied:
Makes sense, I'll do that.

mkouba commented on independent-projects/arc/runtime/src/main/java/io/quarkus/arc/ActiveResult.java:
```diff
@@ -0,0 +1,86 @@
+ * Can say whether the bean {@linkplain #result() is active} and if it is not,
+ * what is the {@linkplain #inactiveReason() reason}.
+ */
+public final class ActiveResult {
```
Maybe use java record?

Ladicek replied:
Public records always have a public canonical constructor with all of their components, which is not what we want here.

mkouba replied:
Why? You can keep the static factory methods and perform the checks in the constructor...

mkouba replied:
Nevermind, it's just a nitpicking comment ;-).

Ladicek replied:
Sure, but why would we want a public constructor in addition to the factory methods?

mkouba replied:
I don't see that as a problem. You would save 3 methods and 3 fields definitions and just add one more way to instantiate the object, with the same semantics :shrug:.

Ladicek replied:
I just don't like the public constructor, but if you insist, I can rewrite this :-)

mkouba replied:
I do not insist at all.

yrodiere reviewed (approved):
Hey @Ladicek ,

First, sorry for the delay. It's been hard to find time for "features and enhancements" in the past few weeks. But I finally managed to get back to it!

Second, my PR #41929 is in good shape now, and I managed to do everything I needed with your current PR. So, as far as I'm concerned, +1 to merge! I added a few comments, and there are a few unaddressed ones from previous reviews, but I'll let you decide if you want to address them now, later, or never :)

Thanks for all this!

Ladicek commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java:
```diff
@@ -1091,6 +1093,55 @@ private void implementCreateForSyntheticBean(ClassCreator beanCreator, BeanInfo
+                        .append("'");
+                msg.append("\nThis bean is injected into:");
+                for (InjectionPointInfo matchingIP : matchingIPs) {
+                    msg.append("\n\t-");
```
Sure, will fix!

Ladicek replied:
Done.

Ladicek commented:
Thanks @yrodiere, I'll take a look at the open comments and will do something about them. I should be able to mark this as ready for review in the coming days.

Ladicek commented:
Done and marked as ready.

Leaving out for the future:

- an `@Active` qualifier (or pseudo-qualifier, because it would have to be special-cased)
- ability to use an exception as a cause of `ActiveResult`

yrodiere commented:
Failure seems unrelated... I'll try a rebase.
@Ladicek can you confirm you're not waiting for another approval specifically? Because in that case, I'd merge as soon as CI passes :)

Ladicek commented:
I was hoping @mkouba and/or @manovotn would take a look :-)

mkouba commented:
> I was hoping @mkouba and/or @manovotn would take a look :-)

Will try tomorrow! ;-)

manovotn commented:
I totally forgot about this one, sorry.
I'll also take a look tomorrow :)

mkouba commented on independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java:
```diff
@@ -1265,11 +1279,16 @@ Builder interceptionProxy(InterceptionProxyInfo interceptionProxy) {
             return this;
         }
 
+        Builder checkActive(Consumer<MethodCreator> checkActiveConsumer) {
```
Even if it's not public it would not hurt to document the purpose of `checkActiveConsumer`. Here or on the `checkActiveConsumer` field. It's not quite obvious.

Ladicek replied:
Sounds fair. I didn't bother, because all the other `Consumer<MethodCreator>` fields/methods are not documented either, but I admit that's a pure excuse for being lazy :-)

mkouba replied:
Well, the purpose of `creatorConsumer` and `destroyerConsumer` is IMO more obvious. This whole "inactive" thing is a bit unusual and special..

Ladicek replied:
Added javadoc to `BeanConfiguratorBase.checkActive()` and `BeanInfo.isActive()`. That is the public API we expose here. (`InjectableBean.isActive()` and `checkActive()` were already documented.)

I didn't add javadoc to `BeanInfo.checkActiveConsumer` or `BeanInfo.Builder.checkActive()`, because that is not a public API.

mkouba replied:
> I didn't add javadoc to BeanInfo.checkActiveConsumer or BeanInfo.Builder.checkActive(), because that is not a public API.

I didn't mean the javadoc but merely some comment that explains what it is used for...

Ladicek replied:
Added a short comment to the `BeanInfo.checkActiveConsumer` field as well.

mkouba commented on extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/StartupBuildSteps.java:
```diff
@@ -179,6 +180,27 @@ private void registerStartupObserver(ObserverRegistrationPhaseBuildItem observer
                     mc.load(bean.getIdentifier()));
+
+            // if the [synthetic] bean is not active and is not injected in a user bean, skip obtaining the instance
+            // this means that an inactive bean that is injected into a user bean will end up with an error
```
_"injected into a user bean"_ is IMO not accurate because the code only filters synthetic beans, i.e. class-based beans from extensions and libraries are considered "user beans".

Ladicek replied:
I think that's expected, actually. The point is that only synthetic beans may be inactive -- all class-based beans (and producer beans) are always active, so if they inject the inactive bean, that should be an error, always.

Ladicek replied:
I'll try to rewrite the comment.

mkouba replied:
My point is that we should not call them "user beans" ;-).

> so if they inject the inactive bean, that should be an error, always.

That's weird. I mean, if an bean is injected into a non-synthetic bean we will initialize the bean during startup as normally and if the bean that injects this inactive bean is never created a user will never know there's a problem?

Ladicek replied:
I guess we could replace `!ip.getTargetBean().get().isSynthetic()` with `!ip.getTargetBean().get().canBeInactive()` and rename `isInjectedInUserBean` to `isInjectedInAlwaysActiveBean`. I guess that would reflect the intent better. We skip initialization here because that bean may also be inactive -- but that only applies to a subset of synthetic beans.

mkouba replied:
> I guess we could replace `!ip.getTargetBean().get().isSynthetic()` with `!ip.getTargetBean().get().canBeInactive()` and rename `isInjectedInUserBean` to `isInjectedInAlwaysActiveBean`. I guess that would reflect the intent better.

:+1: 

> We skip initialization here because that bean may also be inactive -- but that only applies to a subset of synthetic beans.

Ok, so my previous comment was wrong. The check is only perfomed if the bean `A` is NOT injected in a non-synthetic bean. In other words, we only "deactivate" the bean if it's not injected somewhere?

Ladicek replied:
Yeah, sure. If the possibly-inactive synth bean is injected into an always active bean (which are all non-synth beans and some synth beans), we let the eager init happen, which will throw an exception. There's nothing else we can do.

If, however, the possibly-inactive synth bean is only injected into other possibly-inactive synth beans, we skip the eager init, because it might not be needed at all in the end.

Ladicek replied:
Done.

mkouba replied:
> If, however, the possibly-inactive synth bean is only injected into other possibly-inactive synth beans, we skip the eager init, because it might not be needed at all in the end.

This is really weird. `@Startup` is not only about "eager init" of a bean but about executing _any logic_ during startup. Also the bean can be looked up programmatically.

I would naively expect that the synthetic observer for bean `A` (configured with `BeanConfigurator#startup()`) would be notified iff the bean is active. No matter if it is injected anywhere.

Ladicek replied:
Sure, but doing that would mean that we don't throw an exception in case the bean is inactive. Which we don't want. We want, as much as possible, to throw an exception eagerly if an inactive bean is used somewhere.

mkouba replied:
I don't understand. The observer obtains the beans instance first. So it would check the inactivity when the bean instance is created and fail. At least that's what I see in the `BeanGenerator#implementCreateForSyntheticBean()`.

I'm sorry for my ignorance. Just trying to understand the logic and I might be missing something obvious.

Ladicek replied:
That's exactly right. So we let the observer create an instance with the expectation that it will fail.

We only short-circuit that, skipping eager init, if the bean is inactive and is only injected into other possibly inactive beans. In that case, we don't know if we should throw an error eagerly or not. In all other cases, we do know that.

Ladicek replied:
I think this is done, right?

mkouba commented:
Shouldn't we add some docs in the [CDI integration guide](https://quarkus.io/guides/cdi-integration#synthetic_beans)?

Ladicek commented:
Hi @yrodiere, in the conversation with @mkouba (see above), we slightly changed the logic in `StartupBuildSteps.registerStartupObserver()`. Previously, we skipped eager init in case the inactive synthetic bean was injected only into synthetic beans. Now, we skip eager init in case the inactive synthetic bean is injected only into other possibly-inactive synthetic beans. I don't think that should make a difference for you, but could you please check with your branch? Thanks!

mkouba commented on docs/src/main/asciidoc/cdi-integration.adoc:
```diff
@@ -419,6 +419,71 @@ public class TestRecorder {
+If an inactive bean is injected somewhere, or is dynamically looked up, an `InactiveBeanException` is thrown.
+The error message contains the reason (from the `ActiveResult`), the cause chain (also from the `ActiveResult`), and possibly also a list of all injection points that resolve to this bean.
+
+This means that a possibly inactive bean should only be injected using `Instance<\...>`.
```
```suggestion
This means that a possibly inactive bean should only be injected using `Instance<>`.
```

Ladicek replied:
If it's the `\` that looks weird, that's because it prevents Asciidoc from folding 3 periods into an ellipsis.

Ladicek replied:
Here's more discussion about that exact issue: https://stackoverflow.com/questions/65353249/prevent-asciidoctor-from-doing-ellipsis-replacement-in-inline-literal

mkouba replied:
Not really, I usually don't use 3 periods for parameterized types, i.e. I prefer `Instance<>` to `Instance<...>`. But that's just nitpicking..

Ladicek replied:
Fair enough. I vaguely recall seeing `<>` elsewhere. I'll change that to retain consistency.

mkouba commented on docs/src/main/asciidoc/cdi-integration.adoc:
```diff
@@ -419,6 +419,71 @@ public class TestRecorder {
+[source,java]
+----
+@Inject
+InjectableInstance<Foo> foo;
```
We should add `import io.quarkus.arc.InjectableInstance;` so that it's clear where the class comes from.

Ladicek replied:
Good point.

Ladicek replied:
Done.

mkouba commented on docs/src/main/asciidoc/cdi-integration.adoc:
```diff
@@ -419,6 +419,71 @@ public class TestRecorder {
+}
+----
+<1> A bean that might be inactive is typically initialized eagerly, to make sure that an error is thrown at application startup.
+    If the bean is in fact inactive, but is not injected into an always-active bean, eager initialization is skipped and no error is thrown.
```
Maybe this should be mentioned in the javadoc of  `BeanConfiguratorBase#startup()` methods.

Ladicek replied:
I suppose that makes sense, even though it's useless for most people.

Ladicek replied:
Done.

mkouba replied:
> I suppose that makes sense, even though it's useless for most people.

I suppose that this whole feature is kind of useless for "most" people :smile:.

manovotn commented on docs/src/main/asciidoc/cdi-integration.adoc:
```diff
@@ -419,6 +419,73 @@ public class TestRecorder {
+@Inject
+InjectableInstance<Foo> foo;
+
+if (foo.getHandle().getBean().isActive()) {
```
Could we perhaps make our impl of `Instance.isResolvable` aware of this and use that instead?
User would then not need to know the gritty details of bean being possibly inactive and would just use standard APIs.

mkouba replied:
Interesting idea. Note that we already filter beans that are [suppressed](https://quarkus.io/guides/cdi-reference#declaratively-choose-beans-that-can-be-obtained-by-programmatic-lookup). Maybe an inactive bean should be automatically suppressed?

Ladicek replied:
That is a good point!

yrodiere replied:
I need a way to find beans that are inactive, in order to propagate their "inactive reason" to other beans that also become inactive as a result. See [this](https://github.com/quarkusio/quarkus/pull/41929/files#diff-ea1e7589789d9a9f8ee570a12d6cd973cb23cd7aa97f4a14eea1266a110779ccR59-R84) for example:

```java
    public Supplier<ActiveResult> flywayCheckActiveSupplier(String dataSourceName) {
        return new Supplier<ActiveResult>() {
            @Override
            public ActiveResult get() {
                // Flyway beans are inactive when the datasource itself is inactive.
                var dataSourceBean = AgroalDataSourceUtil.dataSourceInstance(dataSourceName).getHandle().getBean();
                var dataSourceActive = dataSourceBean.checkActive();
                if (!dataSourceActive.value()) {
                    return ActiveResult.inactive(
                            String.format(Locale.ROOT,
                                    "Flyway for datasource '%s' was deactivated automatically because this datasource was deactivated.",
                                    dataSourceName),
                            dataSourceActive);
                }

                // Note: When quarkus.flyway.active is set to false, Flyway beans are still available.
                //       The property only controls automatic execution on startup.
                // TODO should we change quarkus.flyway.active (see ^) to align on other extensions?
                //   See https://github.com/quarkusio/quarkus/issues/42244.
                //   We'd have something like quarkus.liquibase.startup.enabled controlling startup behavior,
                //   and *if necessary* quarkus.liquibase.active controlling bean availability
                //   (though IMO controlling that at the datasource level would be enough).
                return ActiveResult.active();
            }
        };
    }
```

If you just filter out inactive beans, will I still be able to find them and get their "inactive reason"?

Or do you want to implement the "inactive bean propagation" directly in Arc? That's tempting, but I can see it having unintended side-effects in the future (deactivating beans where it makes no sense).

manovotn replied:
> If you just filter out inactive beans, will I still be able to find them and get their "inactive reason"?

My initial comment didn't mean to filter beans - we could just alter how `Instance.isResolvable()` work to account for inactive beans (i.e. returning false if it is inactive). You'd still be able to get the metadata.

Either way, with https://github.com/quarkusio/quarkus/pull/41810#discussion_r1789647409 in mind, I am not sure it's worth the hassle :)

Ladicek replied:
So I don't think we can make inactive beans suppressed, because those cannot be obtained via `Instance` at all.

Whether we can make the `Instance` non-resolvable, I'm not sure. Let me think about that.

Ladicek commented on docs/src/main/asciidoc/cdi-integration.adoc:
```diff
@@ -419,6 +419,71 @@ public class TestRecorder {
+If an inactive bean is injected somewhere, or is dynamically looked up, an `InactiveBeanException` is thrown.
+The error message contains the reason (from the `ActiveResult`), the cause chain (also from the `ActiveResult`), and possibly also a list of all injection points that resolve to this bean.
+
+This means that a possibly inactive bean should only be injected using `Instance<\...>`.
```
We could do that during build, yes. If someone directly injects a bean that is possibly-inactive, that is probably an error.

Currently, we don't -- we only do that at runtime.

manovotn replied:
Personally, I think we should enforce it.
ATM we are saying that people should do that anyway and we assume they need to check if the bean is active - both parts can be part of an error msg user gets if they try to perform a direct injection of such a bean.

Ladicek replied:
Yeah, agree. I'll try to add a build-time validation.

yrodiere replied:
Wait, datasources are always "possibly inactive". Are we really talking about forbidding users from injecting datasources without wrapping them in `Instance<>`? That seems... extreme?

Please keep in mind that while beans such as datasources are "possibly inactive", in most applications it will be very unlikely:

1. While datasources expose a ".active" runtime configuration property, the vast majority of users won't use it or even know it's there.
2. The only remaining way for a datasource to be inactive is the result of a (runtime) configuration mistake, i.e. forgetting to set the datasource URL.

So I don't think it's a good idea to fail at build time if a "possibly inactive" bean is injected without being wrapped in `Instance<>`.

Ladicek replied:
Ouch. Yeah, you're right, that does seem excessive.

I'll try to highlight this in the documentation. If you want to handle the inactive case gracefully, you should always inject using `Instance<>`. If you want the inactive case to blow up, just use regular injection.

manovotn replied:
Fair enough.
We should then avoid recommending `Instance` as a universal approach.

Ladicek replied:
Fixed the documentation.

Ladicek replied:
I believe this is done now.

Ladicek commented on extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/SyntheticBeanBuildItem.java:
```diff
@@ -92,7 +100,11 @@ public static class ExtendedBeanConfigurator extends BeanConfiguratorBase<Extend
-                        "Synthetic bean does not provide a creation method, use ExtendedBeanConfigurator#creator(), ExtendedBeanConfigurator#supplier(), ExtendedBeanConfigurator#createWith()  or ExtendedBeanConfigurator#runtimeValue()");
+                        "Synthetic bean does not provide a creation method, use ExtendedBeanConfigurator#creator(), ExtendedBeanConfigurator#supplier(), ExtendedBeanConfigurator#createWith() or ExtendedBeanConfigurator#runtimeValue()");
+            }
+            if (checkActive != null && supplier == null && runtimeValue == null && fun == null && runtimeProxy == null) {
```
That is not entirely correct -- it is possible (albeit unlikely) that the caller uses `creator(Consumer<MethodCreator>)` or `creator(Class<? extends BeanCreator>)`.

manovotn replied:
I see. I misunderstood it, resolving conversation.

yrodiere commented:
> Hi @yrodiere, in the conversation with @mkouba (see above), we slightly changed the logic in `StartupBuildSteps.registerStartupObserver()`. Previously, we skipped eager init in case the inactive synthetic bean was injected only into synthetic beans. Now, we skip eager init in case the inactive synthetic bean is injected only into other possibly-inactive synthetic beans. I don't think that should make a difference for you, but could you please check with your branch? Thanks!

Thanks for the heads-up. I don't think it will be a problem, and in fact it's something that I wanted initially, but that we had postponed as not being "MVP-worthy", so I'm glad to see it in :) See #41466: 

> In the future we'll want to trigger initialization on startup even if a bean is only used in other synthetic beans (e.g. a datasource in a Hibernate ORM persistence unit), but that will require more work as we'll want to ignore synthetic consumers that are themselves inactive.

Anyway, I rebased my PR, which is building there: https://github.com/yrodiere/quarkus/actions/runs/11210350803
If that build passes, we're all good.

Ladicek committed:
ArC: fix documentation of InjectableBean.getPriority()

yrodiere commented:
> Anyway, I rebased my PR, which is building there: https://github.com/yrodiere/quarkus/actions/runs/11210350803
> If that build passes, we're all good.

Tests passed :tada:

Ladicek commented:
@mkouba / @manovotn, any last comments before I merge? :grin:
